Show steepness of a route

shakhat opened the pull request:
Related to #7358 

The altitude chart can now show the slope of route parts, e.g.: 
![Screenshot from 2024-09-25 15-50-41](https://github.com/user-attachments/assets/c72e1b20-1bdc-4e0c-9799-26ca928cfa44)

shakhat commented:
Steep parts of a pedestrian route are highlighted with a different color:
![Screenshot_20240929_000622](https://github.com/user-attachments/assets/7acffb2d-2847-46f5-a67d-34cc32584e6c)
![Screenshot_20240929_112151](https://github.com/user-attachments/assets/227e8759-1a7a-4ca7-babb-d8dee33c91d0)

foss- commented:
This is fantastic and a highly welcome change. Been missing this when planning bike trips. Just wondering if it would be useful to use stronger colors like green: flat, yellow: light ascends, red: steep ascends. Also wondering if using those colors for the top line would make sense. I don't have komoot, but would be interesting to see how they implemented this feature, as I remember they had a nice UI when seeing it on a friends device.

shakhat commented:
Now with brighter colors:
![Screenshot_20241004_122630](https://github.com/user-attachments/assets/9dbcd52e-0c90-4675-81f5-770d3ae5d6f9)
![Screenshot_20241004_120631](https://github.com/user-attachments/assets/05b12da1-86b4-4e72-a3d8-cfb495ffba2f)

There are a few things that change the existing behavior:
 1. Dashes are removed mainly because some of highlighted segments may fall into gaps
 2. After the first stop the route is not shown darker anymore (not sure what was intention to draw it darker)

Some potential enhancements:
1. The slope data is quite noisy (ex. it can be a short flat segment in the middle of the long ascend). Seems that smoothing could be a good option to reduce noise and make the data more readable (Komoot seems doing this). 
2. There may be an inconsistency between the route and the chart (reasons is that the chart already applies filtering and calculates slope on re-scaled data).
3. Color coding of slope - probably better to skip yellow and start with orange then red and then bordeaux.

biodranik commented:
How can non-geeks easily understand this color coding (think: your parents/grandparents)? Is it beneficial for the majority of users? Isn't the steepness obvious from the graph itself?

The current elevation graph lacks more important info, for example, the lowest/highest elevation points, the elevation difference, etc.

shakhat commented:
Main motivation is that currently there is no way to understand which part of a path is more difficult (has larger slope). The chart below gives a hint, but 1) it is hard to correlate slope / spike from the chart back to the map and 2) the chart is normalized meaning that a user has to take ascend/descend values into account.

Regarding color scheme: I think the chart plays a role of legend here, so when an user sees red on the map and red on the chart and that red corresponds to a slope then it means there is a slope on a route too. Of course this conclusion may be non-universal since even red color means different in different cultures.

Alternative approach could be to show some marks on a graph (e.g. beginning of an ascend) and show the same on the route. Not sure if it would be easier to understand though.

vng commented:
I like the idea! Will review the code later. Should fix now:
- shaders compilation error on iOS. Do you have a Mac to test on iOS?
- should generate Vulkan shaders: ```ANDROID_HOME=<path_to>/Android/sdk ./tools/unix/generate_vulkan_shaders.sh``` and add them to a separate commit.

I'd like to see a route profile with line colors and trapezia colors (like on your first screen), no? Better to distinguish on long routes:
![image](https://github.com/user-attachments/assets/81c23aaf-f7e8-4a08-9eaa-7527ebe89dcf)

pastk commented:
It'd be a great improvement, thanks for working on it!

Some thoughts/points.

I also prefer when the line itself is colored (i.e. not like a colored histogram). Smoothing out slope changes would be good (could be left for next phase though).

I'm not sure coloring the route line itself is a good idea. Do other apps do it and it looks good?
At the moment the route line has different pattern (solid, short dashed, long dashed) and color dependent on the mode (car, ped, cycling). Its also consistent with the "guide line" used during the navigation and helps a user remember what is the current mode. This "reminder" could be achieved somehow else, but it needs discussion. Also the line changes color after an intermediate stop to make it easier to see different "route legs" (though its not very helpful with many stops). My suggestion would be to focus on the graph itself first and continue discussing route line coloring in the meantime.

(I like an idea to expand usage of icons placed along the route line, i.e. an icon might be added to beginning of very steep sections, stairs, technical/dangerous sections etc. but its a different issue https://github.com/organicmaps/organicmaps/issues/7430)

shakhat commented:
Thanks all for the feedback! It seems this patch is better to narrow down to change the chart only and keep route rendering as it is.

As of route coloring, I think the main challenge is to keep it good readable. Current colors (blue/pink) are contrast on any type of the background, while for example green segments get lost in front of a forest.

DeathfireD commented:
I noticed this PR while writing up a feature proposal and figured I'd chime in since it's related. 

> And for the best experience, we should show the route profile view during Navigation (after pressing Start button). And also show the current position cursor in the profile view.

I second adding the route profile graphic after pressing the Start button for navigation. Maybe have a small representation of it with the current include percent and then display the full profile in the slide up menu, where the stop button is. If possible, also add an indicator on it of where you are on the profile and what the current steepness/incline is in that spot. As a hiker it's great to see this type of information in real time to prepare for whats coming up and determine if I should take a break now or push through the steep part. I imagine it's probably the same for Cyclists. 

See these two tickets:
Indicating gradients when guiding bicycle paths
https://github.com/organicmaps/organicmaps/issues/9200

Hiking Path Difficulty Ratings Based on Route Plotted
https://github.com/organicmaps/organicmaps/issues/9525

biodranik commented:
Isn't steepness obvious from the graph itself, without any additional (and often confusing) coloring?

pastk commented:
> Isn't steepness obvious from the graph itself, without any additional (and often confusing) coloring?

Nope, because the graph could have a very different aspect ratio depending on total distance and altitude difference.
Hence a section of the same e.g. 20% incline could be displayed with a very differently angled line.

pastk commented:
> Thanks all for the feedback! It seems this patch is better to narrow down to change the chart only and keep route rendering as it is.

Any progress?
(just checking - no pressure!)

ecstrema commented:
Hey!

Currently in a bike trip, and the elevation chart has been incredibly useful so far. 

I've been missing a way to drag a slider on the chart to know where the different part of the elevation chart ends up being on the map, but meanwhile, this would do the job. 

My two cents towards the steepness calculation:

For most people, altitude is more important than steepness. I would rather have the elevation colors instead of the steepness profiles. However deciding on a color range is hard and unintuitive. 

Secondly, if we go with steepness, smoothing is a must. A simple moving average, with a window based on the total length would probably do the job very well. 

Thirdly, steepness can also be  expressed on a map layer instead of on the path. This is what ski apps do with a semi transparent steepness layer. In the case of navigation, I don't think it's a viable option, but it's something to think about. 

So, to summarize: we want to be able to:
- know where it is steep
- know the altitude at any point

We currently have none of this: we only have an elevation overview. However, both would be solved with the following abilities:

1) zoom the elevation profile. 
2) steepness coloring on only the elevation profile, if we want it. This can rapidly noisy, so I think it should be disabled by default or tuned so it shows the bare minimum. 
3) See what these correspond to in the map. A draggable slider indicating the equivalent position on the path and on the graph would be incredible.

These three would solve all the use case I can think of. Am I missing some use cases?

Of course these three are not simple to implement, but since they would cover all the use cases, it would probably be worth it. 

I'll try to have a look at this when I am back to my computer.

Timocop commented:
This is a must-have for any E-Scooter driver. My city has some steep slopes while being on the road and can be quite dangerous when getting stuck with on E-Scooter on those. Sadly the current elevation graph isnt very useful to detect slopes at all, but this PR makes it way more useful.

My current workaround without this PR is to check every 1km of my track manually if the elevation exceeds 30m (which is somewhat of a 8% elevation slope), but thats very tedious and not very reliable.

biodranik commented:
@Timocop from your description, you don't need a sophisticated elevation graph, but you need an E-Scooter-friendly route that will automatically avoid slopes of 8% and more, right?

Timocop commented:
@biodranik Indeed, it would be nice to add it to the track difficulty as well and not just as a visual graph (such as OsmAnd's "Avoid steep slopes").
But so far, in my dire need of a proper E-Scooter navigation app, most other apps that ive tested (such as OsmAnd) even when set to avoid steep slopes and prioritize even/flat terrain, still have some inclines about 20% in them for some odd reason even if theres alternative routes near them. They are not made for E-Scooters in mind but rather E-Bikes/Bikes which have no issues with those inclines.
I have yet to see E-Scooter profiles on navigation apps.

But this PR would make it so much easier to see what roads i can avoid without me switching constantly to OsmAnd to check the elevation graph how steep a road is or do the workaround mentioned earlier.

shakhat committed:
Show steepness of a route

Pedestrian and bicycle routes are tinted depending
on segment's slope: the higher the slope the more
shifted the color.

Signed-off-by: Ilya Shakhat <shakhat@gmail.com>

shakhat commented:
In the current version the steepness is visualized as color tint - the higher the slope the color is more shifted to red. The altitude chart is unchanged to keep clean look.

Open items / questions:
 * To add a separate color for bicycle routes as red doesn't differ much from violet
 * Now both going up and down have the same tint, not sure if this needs to be changed as from one side going down looks easier, but with a stroller it may be not

biodranik commented:
- Our expert cyclist users prefer to see gradient colors on the map, not on the graph (it's harder to match). Or on both the graph and the map.
- Most of the users don't need it.
- The planned route on the map doesn't have much valuable info, so coloring it only for cycling mode/routes should be okay.
- We plan to make a focused map style for cycling.

@shakhat do you need help with rebase?
